Source/Core/DiscIO/RiivolutionPatcher.cpp

```diff
@@ -364,148 +364,109 @@ static FSTBuilderNode* FindFileNodeInFST(std::string_view path, std::vector<FSTB
+                                std::vector<DiscIO::FSTBuilderNode>* fst,
+                                DiscIO::FSTBuilderNode* dol_node)
 {
-  for (const auto& child : external_files)
+  if (!file.m_disc.empty() && file.m_disc[0] == '/')
   {
-    const std::string child_disc_path = std::string(disc_path) + "/" + child.m_filename;
-    const std::string child_external_path = external_path + "/" + child.m_filename;
-    if (child.m_is_directory)
-    {
-      if (recursive)
-      {
-        ApplyFolderPatchToFST(patch, folder,
-                              patch.m_file_data_loader->GetFolderContents(child_external_path),
-                              child_external_path, recursive, child_disc_path, fst);
-      }
-    }
-    else
-    {
-      DiscIO::FSTBuilderNode* node = FindFileNodeInFST(child_disc_path, fst, folder.m_create);
-      if (node)
-        ApplyPatchToFile(patch, node, child_external_path, 0, 0, folder.m_length, folder.m_resize);
-    }
+    // If the disc path starts with a / then we should patch that specific disc path.
+    DiscIO::FSTBuilderNode* node =
+        FindFileNodeInFST(std::string_view(file.m_disc).substr(1), fst, file.m_create);
+    if (node)
+      ApplyPatchToFile(patch, file, node);
+  }
+  else if (CaseInsensitiveEquals(file.m_disc, "main.dol"))
+  {
+    // Special case: If the filename is "main.dol", we want to patch the main executable.
+    ApplyPatchToFile(patch, file, dol_node);
+  }
+  else
+  {
+    // Otherwise we want to patch the first file in the FST that matches that filename.
+    DiscIO::FSTBuilderNode* node = FindFilenameNodeInFST(file.m_disc, *fst);
+    if (node)
+      ApplyPatchToFile(patch, file, node);
   }
 }
 
-static void ApplyUnknownFolderPatchToFST(const Patch& patch, const Folder& folder,
-                                         const std::vector<FileDataLoader::Node>& external_files,
-                                         const std::string& external_path, bool recursive,
-                                         std::vector<DiscIO::FSTBuilderNode>* fst)
+static void ApplyFolderPatchToFST(const Patch& patch, const Folder& folder,
+                                  std::vector<DiscIO::FSTBuilderNode>* fst,
+                                  DiscIO::FSTBuilderNode* dol_node, std::string_view disc_path,
+                                  std::string_view external_path)
 {
+  const auto external_files = patch.m_file_data_loader->GetFolderContents(external_path);
   for (const auto& child : external_files)
   {
-    const std::string child_external_path = external_path + "/" + child.m_filename;
+    const auto combine_paths = [](std::string_view a, std::string_view b) {
+      if (a.empty())
+        return std::string(b);
+      if (b.empty())
+        return std::string(a);
+      if (StringEndsWith(a, "/"))
+        a.remove_suffix(1);
+      if (StringBeginsWith(b, "/"))
+        b.remove_prefix(1);
+      return std::string(a) + "/" + std::string(b);
```

Comment:
NBD but I think `fmt::format("{}/{}", a, b)` may be a bit more performant

Reply:
Good point, yes.